gdx/src/com/badlogic/gdx/utils/ArraySupplier.java

```diff
@@ -0,0 +1,12 @@
+
+package com.badlogic.gdx.utils;
+
+/** An interface that is used to create arrays. Even tho not annotated with "FunctionalInterface", it can act as one. <br>
+ * You can use a constructor reference or lambda as a ArraySupplier, such as with {@code MyClass[]::new} or
+ * {@code (size) -> new MyClass[size]}. */
+public interface ArraySupplier<T> {
+	/** A default array supplier that creates an Object[]. */
+	ArraySupplier<?> OBJECT_ARRAY_SUPPLIER = Object[]::new;
```

Comment:
I think I'd rather see `Object[]::new` inline everywhere than `ArraySupplier.OBJECT_ARRAY_SUPPLIER`.

Reply:
The main issue was, that when doing that, you need to do:
```
(ArraySupplier<T[]>)(ArraySupplier<?>)Object[]::new
``` 
, which was in my opinion rather noisy.
But I don't have a strong opinion on that, I can move it to inline.

Reply:
Agreed, it's a lot shorter. As long as `Object[]::new` doesn't capture surrounding state, we should be fine just using the constructor reference.

Reply:
The extra noise would be an issue to the shortness, hm. Honestly, my only opposition to any chosen syntax is based on how much space any of these takes up, since we may see them a lot. The type inference with method and constructor references is something I don't have as much experience with, and it seems rather different from the language-level-7 techniques we were restricted to before.

How about `ArraySupplier.ANY` or `ArraySupplier.OBJECT`? They both say what it can be used for and are significantly less wordy.

Reply:
It could be `(ArraySupplier)Object[]::new`. Raw type usage discards generics safety checks. If we can agree on disabling the right warnings (libgdx has been doing for ~a decade) then we can use this straightforward approach when necessary without warnings, `@SuppressWarnings`, arcane casts, extra fields, etc.

Reply:
How about this?
```java
public interface ArraySupplier<T> {
	/** A default array supplier that creates an Object[]. */
	ArraySupplier<?> ANY = Object[]::new;

	/** Returns a default array supplier that creates an Object[]. */
	@SuppressWarnings("unchecked")
	static <T> ArraySupplier<T[]> object () {
		return (ArraySupplier<T[]>)ANY;
	}

	T get (int size);
}
```

Then we can just do `this(initialSize, ArraySupplier.object());`. No warnings that need surpression and looks not bad imo.

Reply:
`(ArraySupplier<T[]>)(ArraySupplier<?>)` via a field and method reads OK as `ArraySupplier.object()`, but I contend that `(ArraySupplier)Object[]::new` is simplest and the code is clearer. No warnings with the settings required for libgdx to have no warnings. No suppression is needed. Raw types are a tool for exactly this situation, where generics get in the way.

Reply:
> No warnings with the settings required for libgdx to have no warnings. No suppression is needed. Raw types are a tool for exactly this situation, where generics get in the way.

The issue is, that IntelliJ by default issues the warnings. And also javac issues linter warnings for them.
We can look into ways, to disable these warnings for IntelliJ, eclipse and javac, but I think this is out of scope for this PR.
Regardless, I personally like `ArraySupplier.object()` to, because it imo falls in line with other java 8 added API style (e.g. `Function.identity()` and similar).
So, if you are fine with `ArraySupplier.object()`, I would personally prefer that.

Reply:
IMHO when defaults aren't great, we should configure our tools appropriately.

I'm OK with `ArraySupplier.object()` if you feel strongly about it. The call looks nice, though slightly less clear what it does and doesn't handle eg `(ArraySupplier)Object[][]::new`.

The implementation could still use raw types instead of `(ArraySupplier<T[]>)(ArraySupplier<?>)` shenanigans. We already use raw types, there is no point in make the code less clear to avoid them.
```
static <T> ArraySupplier<T[]> object () {
	return (ArraySupplier)Object[]::new;
}
```

Reply:
Is this code even used enough to justify the concerns? Nobody cares about a slightly more verbose call that isn't often used.

People do care about having to configure their ide/javac to ignore warnings though.

Reply:
>Raw types are a tool for exactly this situation, where generics get in the way.

No, they exist for [legacy code](https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.8).

>Raw types can be thought of as wildcards whose type rules are deliberately unsound, to accommodate interaction with legacy code.

Reply:
> No, they exist for [legacy code](https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.8).

Why they exist doesn't make them any less of a tool for exactly this situation.

@Berstanio I'm not sure what situations those are, but possibly a raw type cast would be applicable there.

Reply:
> The use of raw types is allowed only as a concession to compatibility of legacy code. The use of raw types in code written after the introduction of generics into the Java programming language is strongly discouraged. It is possible that future versions of the Java programming language will disallow the use of raw types.

Not a tool, a concession. It literally says "strongly discouraged" and also that future versions can potentially disallow the use of raw types, which doesn't sound like a very good tool.

Reply:
Why they exist doesn't make them any less of a tool for exactly this situation. You guys not read good? 😛

Despite the JLS's portents, it is unlikely to be removed. In the meantime, it will be enjoyed. I refuse to live in fear!

Reply:
> @Berstanio I'm not sure what situations those are, but possibly a raw type cast would be applicable there.

Yes, raw-types would work, but they issue a warning (especially in my own code base) and `ArraySupplier.object()` and `ArraySupplier.ANY` is just nicer to work with then casts, especially in my own code bases.
So, as they have use for internal libGDX and also potentionally user applications, I think it's a convinient addition.

Reply:
So, what is the state of this? Is this mergable as-is? @NathanSweet

Reply:
I prefer the cast, but I don't mind the static `object()`. Merged now!

> but they issue a warning

They do not if you setup your warnings as intended for the libgdx codebase. In the future I would like to remove unnecessary SuppressWarnings.